celery/canvas.py

```diff
@@ -582,22 +553,47 @@ def stamp(self, visitor=None, **headers):
             headers (Dict): Stamps that should be added to headers.
         """
         self.stamp_links(visitor, **headers)
-
-        headers = headers.copy()
+        headers = deepcopy(headers)
+        visitor_headers = None
         if visitor is not None:
-            visitor_headers = visitor.on_signature(self, **headers)
+            visitor_headers = visitor.on_signature(self, **headers) or {}
+        headers = self._stamp_headers(visitor_headers, **headers)
+        return self.set(**headers)
+
+    def _stamp_headers(self, visitor_headers=None, **headers):
+        headers = deepcopy(headers)
+        if visitor_headers is not None:
+            visitor_headers = visitor_headers or {}
             if "stamped_headers" not in visitor_headers:
                 visitor_headers["stamped_headers"] = list(visitor_headers.keys())
-            _merge_dictionaries(headers, visitor_headers)
+            _merge_dictionaries(headers, visitor_headers, aggregate_duplicates=False)
+            headers["stamped_headers"] = list(set(headers["stamped_headers"]))
         else:
-            headers["stamped_headers"] = [header for header in headers.keys() if header not in self.options]
-            _merge_dictionaries(headers, self.options)
+            stamped_headers = [
+                header for header in headers.keys()
+                if all([
+                    header != "stamped_headers",
+                    header in self.options.get("stamped_headers", [])
+                ])
+            ]
+            if "stamped_headers" in headers:
+                headers["stamped_headers"].extend(stamped_headers)
+            else:
+                headers["stamped_headers"] = stamped_headers
+            for stamp in headers.keys():
+                if stamp != "stamped_headers" and stamp not in headers["stamped_headers"]:
+                    headers["stamped_headers"].append(stamp)
+            headers["stamped_headers"] = list(set(headers["stamped_headers"]))
+            _merge_dictionaries(headers, self.options, aggregate_duplicates=False)
 
         # Preserve previous stamped headers
         stamped_headers = set(self.options.get("stamped_headers", []))
-        stamped_headers.update(headers["stamped_headers"])
+        stamped_headers.update(headers.get("stamped_headers", []))
         headers["stamped_headers"] = list(stamped_headers)
-        return self.set(**headers)
+        for previous_header in stamped_headers:
+            if previous_header not in headers and previous_header in self.options:
+                headers[previous_header] = self.options[previous_header]
+        return headers
```

Comment:
I'm not sure I understand your comment, can you please explain 🙏
Also, please notice the diff between `stamp()` and `_stamp_headers()`, the github diff view might make it look confusing, because I'm not fully sure why you say:
> So now we set all headers every time we stamp? Even the non-stamped headers?

And what difference do you see from the previous implementation logic-wise?

Reply:
Please review again @thedrow , I've refactored the code to fix the previous comments so it might be easier to understand now

Reply:
> I guess your comment should explain what happens during replacing that requires us to always set all stamped headers even if they were not touched by the visitor or by the kwargs.

TBH I don't know. The new test cases are checking so many new edge cases that in the case of stamping with task replacement, this happened.

It is not perfect but it will require a lot of time investigating it just for the comment whereas the feature is still not completed, so if it's acceptable, I'd prefer to leave this code with the comment referencing the replace, but not investigate more at the moment as its too expensive for the sake of having a more detailed comment.